nmosnode/aggregator.py

```diff
@@ -83,58 +80,288 @@ def __init__(self, logger=None, mdns_updater=None):
-                        # Re-register
-                        self.logger.writeWarning("404 error on heartbeat. Marking Node for re-registration")
-                        self._registered["registered"] = False
-
-                        if(self._mdns_updater is not None):
-                            self._mdns_updater.inc_P2P_enable_count()
-                    else:
-                        # Client side error. Report this upwards via exception, but don't resend
-                        self.logger.writeError("Unrecoverable error code {} received from Registration API on heartbeat"
-                                               .format(e.status_code))
-                        self._running = False
-                except Exception:
-                    # Re-register
-                    self.logger.writeWarning("Unexpected error on heartbeat. Marking Node for re-registration")
-                    self._registered["registered"] = False
+            if self.aggregator is None:
+                self._discovery_operation()
+            elif self._node_data["node"] and self._node_data["registered"]:
+                self._registered_operation()
             else:
-                self._registered["registered"] = False
-                if(self._mdns_updater is not None):
-                    self._mdns_updater.inc_P2P_enable_count()
-            while heartbeat_wait > 0 and self._running:
-                gevent.sleep(1)
-                heartbeat_wait -= 1
+                self._node_data["registered"] = False
+                self.aggregator = None
+
         self.logger.writeDebug("Stopping heartbeat thread")
 
-    # Provided the Node is believed to be correctly registered, hand off a single request to the SEND method
-    # On client error, clear the resource from the local mirror
-    # On other error, mark Node as unregistered and trigger re-registration
+    def _discovery_operation(self):
+        """In Discovery operation the Node will wait a backoff period if defined to allow aggregators to recover when in
+        a state of error. Selecting the most appropriate aggregator and try to register with it.
+        If a registration fails then another aggregator will be tried."""
+        self.logger.writeDebug("Entering Discovery Mode")
+        # Wait backoff period
+        if not self._aggregator_failure:
+            self._back_off_timer()
+
+        self._aggregator_failure = False
+
+        # Update cached list of aggregators
+        if self._aggregator_list_stale:
+            self._flush_cached_aggregators()
+        while True:
+            self.aggregator = self._get_aggregator()
+            if self.aggregator is None:
+                self.logger.writeDebug("Failed to find registration API")
+                break
+            self.logger.writeDebug("Aggregator set to: {}".format(self.aggregator))
+
+            if self._heartbeat_operation():
+                # Successfully registered Node
```

Comment:
Could you add a comment here that this is an initial heartbeat for confirmation and as such it doesn't set the _aggregator_failure which would normally happen in heartbeat operation? This would help to explain why it differs slightly from the state diagram.

I wonder if the _heartbeat_operation method could be renamed to something that doesn't include 'operation' to avoid confusion of this being called from within _discovery_operation and _registered_operation? Similar updates to the headings in the state diagram would be good too.

Reply:
Renamed to `heartbeat()` and added some comments explaining operation